### ISSUE TITLE
Prefetch mechanism like torch.dataloader in DALI?

### ISSUE THREAD

**Issue by zhiheng-ldj** [NVIDIA/DALI#2467]
Hi,

Thanks for your help!

Does DALI support prefetch during loading data?

I am facing a problem and not sure whether it is caused by prefetch:

I am training a task for detection and here is the log of speed:

~~~
DALI version:
Epoch: [0][0/403]	Time: 1.140s (1.140s)	Speed: 131.5 samples/s	Data: 0.000s (0.000s)	Stage0-heatmaps: 1.410e-05 (1.410e-05)	Stage1-heatmaps: 6.918e-04 (6.918e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 7.436e-20 (7.436e-20)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][100/403]	Time: 1.025s (1.090s)	Speed: 146.3 samples/s	Data: 0.477s (0.577s)	Stage0-heatmaps: 1.101e-04 (1.113e-04)	Stage1-heatmaps: 5.743e-04 (5.880e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 7.698e-10 (3.908e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][100/403]	Time: 1.025s (1.125s)	Speed: 146.3 samples/s	Data: 0.494s (0.567s)	Stage0-heatmaps: 1.100e-04 (1.118e-04)	Stage1-heatmaps: 5.743e-04 (5.880e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.093e-09 (1.401e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][200/403]	Time: 1.018s (1.084s)	Speed: 147.3 samples/s	Data: 0.542s (0.580s)	Stage0-heatmaps: 6.301e-05 (1.009e-04)	Stage1-heatmaps: 5.794e-04 (5.898e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.436e-10 (2.241e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][200/403]	Time: 1.018s (1.102s)	Speed: 147.3 samples/s	Data: 0.507s (0.564s)	Stage0-heatmaps: 6.294e-05 (1.012e-04)	Stage1-heatmaps: 5.791e-04 (5.898e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 4.888e-11 (7.430e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][300/403]	Time: 0.974s (1.094s)	Speed: 154.0 samples/s	Data: 0.462s (0.561s)	Stage0-heatmaps: 1.058e-04 (8.624e-05)	Stage1-heatmaps: 5.885e-04 (5.895e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 2.658e-09 (5.010e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][300/403]	Time: 0.974s (1.083s)	Speed: 154.0 samples/s	Data: 0.495s (0.584s)	Stage0-heatmaps: 1.059e-04 (8.610e-05)	Stage1-heatmaps: 5.900e-04 (5.895e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 4.040e-09 (1.851e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][400/403]	Time: 1.007s (1.090s)	Speed: 149.0 samples/s	Data: 0.500s (0.562s)	Stage0-heatmaps: 9.601e-05 (8.231e-05)	Stage1-heatmaps: 6.931e-04 (5.940e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 5.268e-10 (3.813e-08)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [0][400/403]	Time: 1.006s (1.081s)	Speed: 149.1 samples/s	Data: 0.529s (0.586s)	Stage0-heatmaps: 9.598e-05 (8.220e-05)	Stage1-heatmaps: 6.939e-04 (5.940e-04)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 2.131e-09 (1.831e-08)	Stage1-pull: 0.000e+00 (0.000e+00)
~~~

~~~
torch version:
Epoch: [1][0/452]	Time: 61.242s (61.242s)	Speed: 2.4 samples/s	Data: 58.800s (58.800s)	Stage0-heatmaps: 6.579e-03 (6.579e-03)	Stage1-heatmaps: 1.668e-03 (1.668e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 4.496e-06 (4.496e-06)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][0/452]	Time: 58.966s (58.966s)	Speed: 2.5 samples/s	Data: 56.375s (56.375s)	Stage0-heatmaps: 6.759e-03 (6.759e-03)	Stage1-heatmaps: 1.716e-03 (1.716e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 5.688e-06 (5.688e-06)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][100/452]	Time: 0.629s (1.335s)	Speed: 238.4 samples/s	Data: 0.000s (0.677s)	Stage0-heatmaps: 3.511e-03 (4.823e-03)	Stage1-heatmaps: 1.160e-03 (1.422e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 4.637e-07 (1.339e-06)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][100/452]	Time: 0.630s (1.312s)	Speed: 238.2 samples/s	Data: 0.000s (0.650s)	Stage0-heatmaps: 3.849e-03 (4.833e-03)	Stage1-heatmaps: 1.233e-03 (1.425e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 4.726e-07 (1.247e-06)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][200/452]	Time: 0.631s (1.038s)	Speed: 237.8 samples/s	Data: 0.000s (0.390s)	Stage0-heatmaps: 2.888e-03 (4.054e-03)	Stage1-heatmaps: 1.019e-03 (1.260e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 2.136e-07 (8.567e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][200/452]	Time: 0.631s (1.027s)	Speed: 237.7 samples/s	Data: 0.000s (0.378s)	Stage0-heatmaps: 2.704e-03 (4.058e-03)	Stage1-heatmaps: 9.540e-04 (1.261e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.841e-07 (8.072e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][300/452]	Time: 0.631s (0.933s)	Speed: 237.6 samples/s	Data: 0.000s (0.286s)	Stage0-heatmaps: 2.487e-03 (3.585e-03)	Stage1-heatmaps: 9.038e-04 (1.157e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.771e-07 (6.346e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][300/452]	Time: 0.632s (0.941s)	Speed: 237.5 samples/s	Data: 0.000s (0.293s)	Stage0-heatmaps: 2.362e-03 (3.579e-03)	Stage1-heatmaps: 8.770e-04 (1.157e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.845e-07 (6.656e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][400/452]	Time: 0.627s (0.892s)	Speed: 239.2 samples/s	Data: 0.000s (0.244s)	Stage0-heatmaps: 2.163e-03 (3.261e-03)	Stage1-heatmaps: 8.256e-04 (1.084e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.465e-07 (5.385e-07)	Stage1-pull: 0.000e+00 (0.000e+00)	
Epoch: [1][400/452]	Time: 0.628s (0.887s)	Speed: 238.9 samples/s	Data: 0.000s (0.240s)	Stage0-heatmaps: 2.338e-03 (3.264e-03)	Stage1-heatmaps: 8.712e-04 (1.084e-03)	Stage0-push: 0.000e+00 (0.000e+00)	Stage1-push: 0.000e+00 (0.000e+00)	Stage0-pull: 1.601e-07 (5.134e-07)	Stage1-pull: 0.000e+00 (0.000e+00)
~~~

**Comment by JanuszL** [comment]
Hi,
Yes, DALI does the prefetching and it should be enabled in the default configuration.
Your processing speed may very much depend on your pipeline configuration, if you are using an external source of python operators which are mostly meant for debugging/prototyping, how many threads you have assigned to your pipeline if you are processing using CPU or GPU operators variant?

**Comment by zhiheng-ldj** [comment]
> Hi,
> Yes, DALI does the prefetching and it should be enabled in the default configuration.
> Your processing speed may very much depend on your pipeline configuration, if you are using an external source of python operators which are mostly meant for debugging/prototyping, how many threads you have assigned to your pipeline if you are processing using CPU or GPU operators variant?

Hi, I am using 12 workers (threads) in my setting.

**Comment by JanuszL** [comment]
So the bottleneck is somewhere else. If you can provide a brief idea of how your pipeline looks like it will allow us to tell us more.

**Comment by zhiheng-ldj** [comment]
> So the bottleneck is somewhere else. If you can provide a brief idea of how your pipeline looks like it will allow us to tell us more.

Hi, yes.

~~~
eii = ExternalInputIterator(batch_size=images_per_batch, num_gpus=len(cfg.GPUS), external_data=dataset)
~~~

eii is defined as an External Input Iterator and dataset is a object returning (images_path, targets, masks, joints).

~~~
train_pipes = [ExternalSourcePipeline(batch_size=self.batch_size, num_threads=self.workers, device_id = i, num_gpus = self.world_size, external_data=self.external, transforms=self.transforms) for i in range(self.world_size)]
~~~

I passed eii to ExternalSourcePipeline and build.

**Comment by JanuszL** [comment]
This could be it, ExternalSource doesn't scale with the number of CPU threads passed to the pipeline. We are working to change that, but we cannot provide any timeline yet.

**Comment by zhiheng-ldj** [comment]
> This could be it, ExternalSource doesn't scale with the number of CPU threads passed to the pipeline. We are working to change that, but we cannot provide any timeline yet.

Do you mean now the ExternalSource has the same performance with thread=1?

**Comment by JanuszL** [comment]
> Do you mean now the ExternalSource has the same performance with thread=1?

Unfortunately yes. If possible use build-in readers.

**Comment by zhiheng-ldj** [comment]
> > Do you mean now the ExternalSource has the same performance with thread=1?
> 
> Unfortunately yes. If possible use build-in readers.

Hi,

Thanks for your help. So what should I do to make my code supports multi-workers? Should I replace the ExternalPipeline or just use FileReader? Can I use ops.External source when I define the graph?

Thanks!

**Comment by JanuszL** [comment]
> Should I replace the ExternalPipeline or just use FileReader? 

Please do if possible. you can check [the nightly build](https://docs.nvidia.com/deeplearning/dali/master-user-guide/docs/installation.html#nightly-and-weekly-release-channels) [FileReader](https://docs.nvidia.com/deeplearning/dali/master-user-guide/docs/supported_ops.html#nvidia.dali.ops.FileReader) capabilities, especially:
```
files (str or list of str, optional) –

A list of file paths to read the data from.

If file_root is provided, the paths are treated as being relative to it. When using files, the labels are taken from labels argument or, if it was not supplied, contain indices at which given file appeared in the files list.

This argument is mutually exclusive with file_list.
```
where you can pass a list of file names directly from python.

**Comment by zhiheng-ldj** [comment]
> > Should I replace the ExternalPipeline or just use FileReader?
> 
> Please do if possible. you can check [the nightly build](https://docs.nvidia.com/deeplearning/dali/master-user-guide/docs/installation.html#nightly-and-weekly-release-channels) [FileReader](https://docs.nvidia.com/deeplearning/dali/master-user-guide/docs/supported_ops.html#nvidia.dali.ops.FileReader) capabilities, especially:
> 
> ```
> files (str or list of str, optional) –
> 
> A list of file paths to read the data from.
> 
> If file_root is provided, the paths are treated as being relative to it. When using files, the labels are taken from labels argument or, if it was not supplied, contain indices at which given file appeared in the files list.
> 
> This argument is mutually exclusive with file_list.
> ```
> 
> where you can pass a list of file names directly from python.

Thanks for your help. I am trying. Btw, can I still use ops.ExternalSource as my label? If I do, does it support multiple workers?

**Comment by JanuszL** [comment]
Hi,
As I wrote before - ops.ExternalSource doesn't utilize any mutli-process/threading. You can use it but the performance may be inferior compared to the native readers.

**Comment by JanuszL** [comment]
Hi @zhiheng-ldj,

You can check the most recent improvements in [the `external_source`` oeprator](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/supported_ops.html#nvidia.dali.fn.external_source), especially `parallel` and `prefetch_queue_depth` options. If you use them you should see an improvement in the performance.
